### PR TITLE
Add options to tracked functions for cycle recovery

### LINKED ISSUE #331: Add options to tracked functions for cycle recovery
Salsa allows tracked functions to declare a "cycle recovery" function, [as described here](https://salsa-rs.github.io/salsa/cycles/fallback.html) (actually, that page is somewhat out of date, we no longer require all cycle participants to define a recovery function, for example, should fix that --ed.). This capability is not exposed yet for tracked functions. 

We should support users writing something like

```rust
#[salsa::tracked(recovery_fn = my_recover_fn)]
fn my_tracked_fn(db: &dyn crate::Db, ...) { }

fn my_recover_fn(
    db: &dyn MyDatabase,
    cycle: &salsa::Cycle,
) -> MyResultValue
```

### PR DESCRIPTION
**Pull request by XFFXFF**
closes #331 

This pr ports the old salsa tests for cycle in a single thread, except for [cycle_disappears_durability](https://github.com/salsa-rs/salsa/blob/03a27a7054ee234a566231243dd5025edfbcdde6/tests/cycles.rs#L326), since we don't have the api that permits setting durability.  

~I haven't ported parallel related tests, which would be some work, wondering if we can merge this in first~

### CODE REVIEW

File: salsa-2022-tests/tests/cycles.rs [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+    panic!("unexpected value: {:?}", v)
+}
+
+#[test]
+fn cycle_memoized() {
+    let mut db = Database::default();
+    let input = MyInput::new(&mut db);
+    let cycle = extract_cycle(|| memoized_a(&db, input));
+    let expected = expect![[r#"
+        [
```
Comment (XFFXFF) [inline salsa-2022-tests/tests/cycles.rs:190]: We haven't implemented `debug` for [`DependencyKey`](https://github.com/salsa-rs/salsa/blob/03a27a7054ee234a566231243dd5025edfbcdde6/components/salsa-2022/src/key.rs#L37)

**Comment by nikomatsakis** [comment]
> I haven't ported parallel related tests, which would be some work, wondering if we can merge this in first

yes, let's do that, but let's also create issues for the remaining to-do items...

**Comment by XFFXFF** [comment]
@nikomatsakis  I ported all old salsa tests for cycle recovery, except for [cycle_disappears_durability](https://github.com/salsa-rs/salsa/blob/03a27a7054ee234a566231243dd5025edfbcdde6/tests/cycles.rs#L326), I added a [issue](#339)  for this.  
I also added #340 for more human readble output from  `DependencyIndex.debug`

**Review by nikomatsakis — APPROVED** [review]
This is awesome, thanks!

**Comment by nikomatsakis** [comment]
Egads.

bors r+

**Pull request merged**